v3/src/models/formula/formula-manager.ts

```diff
@@ -307,6 +303,10 @@ export class FormulaManager {
     const { dataSet } = formulaContext
     const { formula, adapter } = formulaMetadata
     const { defaultArgument } = extraMetadata
+
+    // unregister formulas when they're destroyed
+    addDisposer(formula, () => this.unregisterFormula(formulaId))
+
```

Comment:
I could see that in the `FormulaManager#registerFormula` method as well, to keep the register and unregister actions together (kinda like useEffect pattern), but it's also true that it's an observer. :)

Reply:
I think you're right -- `registerFormula` would make more sense. Feel free to make that change in a subsequent PR.